### PR TITLE
[SPARK-43097][ML] New pyspark ML logistic regression estimator implemented on top of distributor

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

This PR takes over https://github.com/apache/spark/pull/40748

Closes https://github.com/apache/spark/pull/40748

### Why are the changes needed?

Distributed ML on spark connect project.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Unit tests.

Manually testing code:

Start local pyspark shell by:

```
bin/pyspark --remote "local[2]"    # spark connect mode
```
to launch pyspark shell,

Then, paste following code to pyspark shell:
```
from pyspark.mlv2.classification import LogisticRegression as LORV2
lorv2 = LORV2(maxIter=2, numTrainWorkers=2)

from pyspark.ml.linalg import Vectors

df = spark.createDataFrame(
    [
        (1.0, [0.0, 5.0]),
        (0.0, [1.0, 2.0]),
        (1.0, [2.0, 1.0]),
        (0.0, [3.0, 3.0]),
    ] * 100,
    ["label", "features"],
)

model.transform(df).show(truncate=False)
model.transform(df.toPandas())

model.set(model.probabilityCol, "")
model.transform(df).show(truncate=False)
model.transform(df.toPandas())
```

### CODE REVIEW

File: python/pyspark/mlv2/classification.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            batchSize=batchSize,
+            learningRate=learningRate,
+            momentum=momentum,
+            seed=seed,
+        )
+
+    def _fit(self, dataset: Union[DataFrame, pd.DataFrame]) -> "LogisticRegressionModel":
+        if isinstance(dataset, pd.DataFrame):
+            # TODO: support pandas dataframe fitting
+            raise NotImplementedError("Fitting pandas dataframe is not supported yet.")
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/classification.py:192]: should use `PySparkNotImplementedError` instead

File: python/pyspark/mlv2/classification.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        # TODO: check label values are in range of [0, num_classes)
+        num_rows, num_classes = dataset.agg(count(lit(1)), countDistinct(self.getLabelCol())).head()
+
+        num_batches_per_worker = math.ceil(num_rows / num_train_workers / batch_size)
+        num_samples_per_worker = num_batches_per_worker * batch_size
+
+        num_features = len(dataset.select(self.getFeaturesCol()).head()[0])
+
+        if num_classes < 2:
+            raise ValueError("Training dataset distinct labels must >= 2.")
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/classification.py:212]: `PySparkValueError` https://github.com/apache/spark/blob/e7a466ebce780664e7601f28b6dab02db4703871/python/pyspark/errors/exceptions/base.py#L195

File: python/pyspark/mlv2/classification.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        model_state_dict = self.torch_model.state_dict()
+        num_features = self.num_features
+        num_classes = self.num_classes
+        fit_intercept = self.getFitIntercept()
+
+        def transform_fn(input_series: "pd.Series") -> Any:
+            torch_model = _LinearNet(
+                num_features=num_features, num_classes=num_classes, bias=fit_intercept
+            )
+            # TODO: Use spark broadast for `model_state_dict`,
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/classification.py:286]: I guess we don't have plan to support broadcast

File: python/pyspark/mlv2/common_params.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from pyspark.ml.param import Param, Params, TypeConverters
+
+
+class HasNumTrainWorkers(Params):
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/common_params.py:21]: what about moving new params to `pyspark.ml.param.shared` ?

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/mlv2/classification.py (hunk 3 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/classification.py:286]: OK. Not a big issue for now.

File: python/pyspark/mlv2/common_params.py (hunk 4 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/common_params.py:21]: Good idea!

File: python/pyspark/mlv2/classification.py (hunk 1 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/classification.py:187]: I think `PySparkXXXError` is only used in pyspark core code.

File: python/pyspark/mlv2/classification.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    )
+
+    loss_fn = torch_nn.CrossEntropyLoss()
+
+    optimizer = optim.SGD(ddp_model.parameters(), lr=learning_rate, momentum=momentum)
+    data_loader = _get_spark_partition_data_loader(
+        num_samples_per_worker,
+        batch_size,
+        num_workers=0,
+        prefetch_factor=None,
```
Comment (HyukjinKwon) [inline python/pyspark/mlv2/classification.py:115]: @WeichenXu123 is this code being tested? It cannot be `None`, and throw an exception like:

```
ValueError: prefetch_factor option could only be specified in multiprocessing.let num_workers > 0 to enable multiprocessing.
```

Comment (WeichenXu123) [reply python/pyspark/mlv2/classification.py:115]: Strange, CI tests and my local tests all passed ?

Comment (WeichenXu123) [reply python/pyspark/mlv2/classification.py:115]: OK this only works for torch>=2.0, I will fix it.

**Pull request closed without merging**